### ISSUE TITLE
[2.7] ACL - Inherit or not?

### ISSUE THREAD

**Issue by davidsteinsland** [cakephp/cakephp#8114]
Given the path "controllers/Hello/world", `Permission::check()` splits the ACOs up and searches bottom-up (with action `*`), so far so good.

Now, let's imagine the following permissions are added:
```
+-----+-------+--------+--------+------+--------+--------+
| id  | aco   |  aro   | create | read | update | delete |
+-----+-------+--------+--------+------+--------+--------+
|  1  | world | User.1 |   1    |   0  |   0    |    0   |
+-----+-------+--------+--------+------+--------+--------+
|  2  | hello | User.1 |   -1   |   0  |   0    |    0   |
+-----+-------+--------+--------+------+--------+--------+
```

In other words: We deny `User.1` `create` access for `controllers/Hello` as a whole, but then we explicitly give `create` access for `controllers/Hello/world`.

One would suspect that the user is given access, right?
Nope – `Permissions::check()` returns false.
Did `controllers/Hello/world` just inherit the denial of `create` for `controllers/Hello`? Yes, it did.

Why?
Because `ActionsAuthorize` does not care about `CRUD`. It searches with action being `*`, which apparently means `all CRUD actions must be allowed`. But wait: `ActionsAuthorize` *does* care about `CRUD`, then?

By looking at `Permissions::check()`, we start to see [the reason why](https://github.com/cakephp/cakephp/blob/2.7/lib/Cake/Model/Permission.php#L135-L139). 

As soon as a `-1` value is encountered, the function returns `false`, but if the value is `1` it gets stored in an array called `$inherited` (even more confusing: isn't `inherit` the value of `0`? Access granted is `1` …).

By looking at the code, we see that:
- For a granted permission `P` over an action `Q`, every possible permission in `P` must evaluate to `1`. In other words, for the function to return `true`, `_create`, `_read`(and so on) must all be `1`.

How does inheriting works, then? If we set `_read`, `_update` and `_delete` to `1` in the last ACO, one would expect that:
- `_create` is denied for `Hello` whilst the other are granted
- `_create` is explicitly granted for `world` whilst the other are inherited

This means that all permissions should evaluate to `1` for `controllers/Hello/world`.
Short answer: it does not.

---

My motivation for this is that I'd like to use the `ActionsAuthorize` to grant access to controllers/actions. Further down the line, I want to specify permissions further using a manual lookup, e.g.:
```php
if (Acl->check($user, 'controllers/Hello/world', '_delete')) {
    // display delete link
}
```

However, this is not possible, as described.

---

Suggestions: 
- Fix properly inheriting
- Make `ActionsAuthorize` search for *any* permission being `1`, and allow access. No need for all `CRUD` to be 1, because then we could just use `CrudAuthorize`.

---

TL;DR The `Permission::check()` function is buggy, poorly documented and suffers from years of neglect and bad testing.

**Comment by markstory** [comment]
Its been a long time since I've looked at DbAcl, but from looking at the tests the intent around ActionsAuthorize is to deny if there could be any permission that would result in a deny. This mirrors how the '*' permission is allowed/granted.

> Make ActionsAuthorize search for any permission being 1, and allow access. No need for all CRUD to be 1, because then we could just use CrudAuthorize

This seems dangerous. While I agree that the current behavior is not entirely rational, there could exist many applications relying on the illogical nuances it contains. At this point in the 2.x development I'd prefer to not break those applications.

**Comment by davidsteinsland** [comment]
Hmm... In my world `*` means "any". I.e. "if any permissions are 1, then allow". In Permissions, it means the same as `CRUD`? I.e. all permissions must be 1?

What I fail to understand is *why* ActionsAuthorize needs all CRUD to be 1. By only needing *one* permission, it's up to the user to take action on what permission that is. It would act as a very simple CrudAuthorize, because:

- I would allow `read` to all my routes
- Some have `delete = 1` instead

ActionsAuthorize would allow all kind of requests, but it's then up to me to check the value of `Permission->check()` and allow delete or read, depending on what the route/action is doing.

Because ... I thought that CrudAuthorize used `REQUEST_METHOD` to determine what kind of action is was (`POST = create, GET = read, ....), but this is not the case: it's up to me to specify a map of `routes --> permission`. So how does CrudAuthorize differs from ActionsAuthorize then? 

Personally, I don't think there are many people that's dependent on the "bogus functionality" you described. This is because `./Console/cake acl grant ARO ACO` sets all `CRUD` to `1` when allowing, which is probably the method people most uses to set permissions.

However, I could agree on a method that focuses on BC also, that you would switch via a configuration:

```php
public $components = [
  'Auth' => array(
    'authorize' => [
      'Actions' => [
        'actionPath' => 'controllers/',
        'permission' => 'any' /* <any>, <all> */
      ]
    ]
  ]
]
```

**Comment by davidsteinsland** [comment]
To fix inherit search, here's a solution:

```php
/* Permission::check() */
$inherited = array_fill_keys($permKeys, 0);
// .....

					foreach ($permKeys as $key) {
						$p = (int)$perm[$key];
						if ($p === -1 || $p === 1) {
							if ($inherited[$key] !== 0) {
								continue;
							}

							if ($p === -1) {
								return false;
							}

							$inherited[$key] = $p;
						}
					}

					$accessGrantedCount = 0;
					foreach ($permKeys as $p) {
						if ($inherited[$p] === 1) {
							$accessGrantedCount++;
						}
					}

					if ($accessGrantedCount === count($permKeys)) {
						return true;
					}
```

Basically, we don't overwrite information in the `inherited` array as you do today, because this will break inherited permissions.

**Comment by markstory** [comment]
> Personally, I don't think there are many people that's dependent on the "bogus functionality" you described. 

I'm not willing to take the chance in a bug fix release sorry.

> So how does CrudAuthorize differs from ActionsAuthorize then? 

You can map actions to specific permissions in CrudAuthorize where as ActionsAuthorize always uses '*'.

**Comment by davidsteinsland** [comment]
Bogus functionality aside – there is also the case of buggy inherit search. This is fixed in code listing in previous comment.

You can test it very quickly with:
```
+-----+-------+--------+--------+------+--------+--------+
| id  | aco   |  aro   | create | read | update | delete |
+-----+-------+--------+--------+------+--------+--------+
|  1  | world | User.1 |   1    |   0  |   0    |    0   |
+-----+-------+--------+--------+------+--------+--------+
|  2  | hello | User.1 |   -1   |   1  |   1    |    1   |
+-----+-------+--------+--------+------+--------+--------+
```
`Permission::check(User.1, hello/world, '*')` should return `true`. Today it doesn't.

**Comment by markstory** [comment]
Thanks, I'll take a look into reproducing the issue with inherits.

**Comment by markstory** [comment]
Pull request up now.
